### PR TITLE
Refactored and improved API docs for DeltaTable operations

### PR DESCRIPTION
**Pull request by tdas**
- Moved the update() methods to DeltaTable class to fix java doc issues
- Converted builder classes from case class to simple class because case classes have unnecessary public methods (e.g. productArity, etc.) that show up in the API docs.
- Added details docs for update and merge

### CODE REVIEW

File: src/main/scala/io/delta/DeltaMergeBuilder.scala [hunk 1]
```diff
@@ -80,36 +252,111 @@ case class DeltaMergeBuilder private[delta](
...
     private val mergeBuilder: DeltaMergeBuilder,
     private val matchCondition: Option[Column]) {
 
+  /**
+   * :: Evolving ::
+   *
+   * Update the matched table row based on the rules defined by `set`.
+   *
+   * @param set rules to update a row as a Scala map between target column names and
+   *            corresponding expressions as Column objects.
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaMergeBuilder.scala]: rules to update a row as a Scala map between target column names and
their corresponding update expressions as Column objects.

File: src/main/scala/io/delta/DeltaMergeBuilder.scala [hunk 2]
```diff
@@ -80,36 +252,111 @@ case class DeltaMergeBuilder private[delta](
...
+ */
+@Evolving
+class DeltaMergeMatchedActionBuilder private(
     private val mergeBuilder: DeltaMergeBuilder,
     private val matchCondition: Option[Column]) {
 
+  /**
+   * :: Evolving ::
+   *
+   * Update the matched table row based on the rules defined by `set`.
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaMergeBuilder.scala]: row -> rows

File: src/main/scala/io/delta/DeltaMergeBuilder.scala (hunk 1 above)
Comment (liwensun) [reply src/main/scala/io/delta/DeltaMergeBuilder.scala]: The above wording suggestions apply to the overloading APIs below.

File: src/main/scala/io/delta/DeltaMergeBuilder.scala [hunk 3]
```diff
@@ -132,31 +379,107 @@ case class DeltaMergeMatchedActionBuilder private[delta](
...
+  }
+}
+
+
+/**
+ * :: Evolving ::
+ *
+ * Builder class to specify the actions to perform when a source row has not matched any target
+ * Delta table row based on the merge condition, but has matched any additional condition
+ * if specified.
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaMergeBuilder.scala]: Builder class to specify the actions to perform when a target row does not match any source row based on the merge condition but matches the `notMatchedCondition` when specified.

File: src/main/scala/io/delta/DeltaMergeBuilder.scala [hunk 4]
```diff
@@ -132,31 +379,107 @@ case class DeltaMergeMatchedActionBuilder private[delta](
...
+   * @since 0.3.0
+   */
+  @Evolving
   def insertAll(): DeltaMergeBuilder = {
     val insertClause = MergeIntoInsertClause(
-      matchCondition.map(_.expr),
+      notMatchCondition.map(_.expr),
       MergeIntoClause.toActions(Nil, Nil))
-    mergeBuilder.copy(whenClauses = mergeBuilder.whenClauses :+ insertClause)
+    mergeBuilder.withClause(insertClause)
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaMergeBuilder.scala]: what is this change about?

File: src/main/scala/io/delta/DeltaMergeBuilder.scala [hunk 5]
```diff
@@ -80,36 +252,111 @@ case class DeltaMergeBuilder private[delta](
...
+      source: DataFrame,
+      onCondition: Column): DeltaMergeBuilder = {
+    new DeltaMergeBuilder(targetTable, source, onCondition, Nil)
+  }
+}
+
+/**
+ * :: Evolving ::
+ *
+ * Builder class to specify the actions to perform when a target table row has matched a
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaMergeBuilder.scala]: has matched -> matches

File: src/main/scala/io/delta/DeltaMergeBuilder.scala [hunk 6]
```diff
@@ -132,31 +379,107 @@ case class DeltaMergeMatchedActionBuilder private[delta](
...
+   */
+  @Evolving
   def insert(values: Map[String, Column]): DeltaMergeBuilder = {
     addInsertClause(values)
   }
 
+  /**
+   * :: Evolving ::
+   *
+   * Insert a new target Delta table row based on the rules defined by `values`.
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaMergeBuilder.scala]: Insert new rows to the target table based on the rules defined by `values`.

apply to all overloaders below.

File: src/main/scala/io/delta/DeltaMergeBuilder.scala [hunk 7]
```diff
@@ -132,31 +379,107 @@ case class DeltaMergeMatchedActionBuilder private[delta](
...
+   *
+   * @param values rules to insert a row as a Scala map between target column names and
+   *               corresponding expressions as SQL formatted strings.
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaMergeBuilder.scala]: and __their__ corresponding expressions

apply to all below.

File: src/main/scala/io/delta/DeltaTable.scala [hunk 8]
```diff
@@ -97,15 +99,248 @@ class DeltaTable (df: Dataset[Row]) extends DeltaTableOperations {
     executeDelete(None)
   }
 
+
   /**
    * :: Evolving ::
    *
-   * Merge data from the `source` table that match the given `condition`
+   * Update data from the table based on the rules defined by `set`.
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaTable.scala]: Update rows in the target able based on the rules defined by `set`.

Comment (liwensun) [reply src/main/scala/io/delta/DeltaTable.scala]: apply to all below.

File: src/main/scala/io/delta/DeltaTable.scala [hunk 9]
```diff
@@ -97,15 +99,248 @@ class DeltaTable (df: Dataset[Row]) extends DeltaTableOperations {
...
+   *
+   *    deltaTable.update(
+   *      new HashMap<String, Column>() {{
+   *        put("data", functions.col("data").plus(1));
+   *      }}
+   *    );
+   * }}}
+   *
+   * @param set rules to update a row as a Java map between target column names and
+   *            corresponding expressions as Column objects.
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaTable.scala]: between target column names and their corresponding update expressions as Column objects

File: src/main/scala/io/delta/DeltaTable.scala [hunk 10]
```diff
@@ -97,15 +99,248 @@ class DeltaTable (df: Dataset[Row]) extends DeltaTableOperations {
...
+  }
+
+  /**
+   * :: Evolving ::
+   *
+   * Merge data from the `source` DataFrame based on the given merge `condition`. This class returns
+   * a [[DeltaMergeBuilder]] object that can be used to specify the update, delete, or insert
+   * actions to be performed on rows based on whether the rows matched the condition or not.
+   *
+   * See the [[DeltaMergeBuilder]] for a full description of all this operation and what combination
```
Comment (liwensun) [inline src/main/scala/io/delta/DeltaTable.scala]: full description of this operation and what combinations of update, ...

Comment (liwensun) [reply src/main/scala/io/delta/DeltaTable.scala]: also below.

File: src/main/scala/io/delta/DeltaMergeBuilder.scala (hunk 4 above)
Comment (tdas) [reply src/main/scala/io/delta/DeltaMergeBuilder.scala]: because i changed from case class to simple class, there were no more copy() methods.

**Pull request closed without merging**